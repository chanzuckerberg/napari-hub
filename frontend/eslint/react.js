module.exports = {
  extends: ['airbnb/hooks'],

  plugins: ['napari-hub'],

  rules: {
    'napari-hub/no-root-mui-import': 'error',

    /*
      Prop types aren't necessary since we have TypeScript interfaces for
      prop types.
    */
    'react/prop-types': 'off',

    /*
      React automatically adds the import as part of the new JSX transform:
      https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#nextjs
     */
    'react/react-in-jsx-scope': 'off',

    /*
      Prop spreading is very useful as a simple way for passing multiple
      props to a component. However, we should still be careful about passing
      unnecessary props.
    */
    'react/jsx-props-no-spreading': 'off',

    /*
      This rule isn't really needed anymore since we can use ES6 object defaults
      for props. There are also plans to deprecate the use of `defaultProps`:
      https://github.com/reactjs/rfcs/pull/107
    */
    'react/require-default-props': 'off',
  },
};
