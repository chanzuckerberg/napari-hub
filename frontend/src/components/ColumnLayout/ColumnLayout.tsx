import clsx from 'clsx';
import {
  createElement,
  HTMLProps,
  MutableRefObject,
  ReactHTML,
  ReactNode,
} from 'react';

interface Classes {
  gap: string;
  twoColumn: string;
  threeColumn: string;
  fourColumn: string;
  fiveColumn: string;
}

type HTMLKey = keyof ReactHTML;

interface Props<T extends HTMLKey, Ref> extends HTMLProps<ReactHTML[T]> {
  /**
   * Children to render as columns. The 1st, 2nd, and 3rd child nodes are used
   * for the 1st, 2nd, and 3rd columns, respectively.
   */
  children: ReactNode;

  /**
   * CSS classes to override component classes.
   */
  classes?: Partial<Classes>;

  /**
   * Root component to use for column layout. The default is `div`.
   */
  component?: T;

  innerRef?: MutableRefObject<Ref>;
}

const defaultClasses: Classes = {
  gap: 'gap-6 screen-495:gap-12',
  twoColumn: 'grid-cols-2',
  threeColumn: 'screen-875:grid-cols-napari-3',
  fourColumn: 'screen-1150:grid-cols-napari-4',
  fiveColumn: 'screen-1425:grid-cols-napari-5',
};

/**
 * Component for rendering napari's column layout.
 */
export function ColumnLayout<T extends HTMLKey, Ref = never>({
  children,
  classes = {},
  className,
  component,
  innerRef,
  ...props
}: Props<T, Ref>) {
  // Use `createElement()` to dynamically create element from `component` prop.
  return createElement(
    component ?? 'div',
    {
      className: clsx(
        className,

        // Grid Layout
        'grid justify-center',

        Object.values({
          // Default classes for component.
          ...defaultClasses,

          // Override classes if any.
          ...classes,
        }),
      ),
      ref: innerRef,
      ...props,
    },
    children,
  );
}
