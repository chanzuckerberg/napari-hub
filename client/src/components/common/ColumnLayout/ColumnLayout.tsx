import clsx from 'clsx';
import { createElement, HTMLProps, ReactHTML, ReactNode } from 'react';

import { NonPartial } from '@/types';

interface Classes {
  gap?: string;
  twoColumn?: string;
  threeColumn?: string;
  fourColumn?: string;
  fiveColumn?: string;
}

type HTMLKey = keyof ReactHTML;

interface Props<T extends HTMLKey> extends HTMLProps<ReactHTML[T]> {
  /**
   * Children to render as columns. The 1st, 2nd, and 3rd child nodes are used
   * for the 1st, 2nd, and 3rd columns, respectively.
   */
  children: ReactNode;

  /**
   * CSS classes to override component classes.
   */
  classes?: Classes;

  /**
   * Root component to use for column layout. The default is `div`.
   */
  component?: T;
}

const defaultClasses: NonPartial<Classes> = {
  gap: 'gap-6 md:gap-12',
  twoColumn: 'grid-cols-2',
  threeColumn: 'screen-875:grid-cols-napari-3',
  fourColumn: 'screen-1150:grid-cols-napari-4',
  fiveColumn: 'screen-1425:grid-cols-napari-5',
};

/**
 * Component for rendering napari's column layout.
 */
export function ColumnLayout<T extends HTMLKey>({
  children,
  classes = {},
  className,
  component,
  ...props
}: Props<T>) {
  // Use `createElement()` to dynamically create element from `component` prop.
  return createElement(
    component ?? 'div',
    {
      className: clsx(
        className,

        // Grid Layout
        'grid justify-center',

        Object.values({
          // Default classes for component.
          ...defaultClasses,

          // Override classes if any.
          ...classes,
        }),
      ),
      ...props,
    },
    children,
  );
}
