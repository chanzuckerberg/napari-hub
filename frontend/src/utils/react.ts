import { ReactElement, ReactNode } from 'react';

/**
 * Checks if a React node is an element. An element is an instantiated
 * component.
 *
 * @param node The React node
 * @returns True if node is an element
 */
export function isReactElement(node: ReactNode): node is ReactElement {
  const element = node as ReactElement | null | undefined;
  return !!(element?.type && element?.props);
}
