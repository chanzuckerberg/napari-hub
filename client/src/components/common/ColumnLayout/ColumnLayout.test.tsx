import { render } from '@testing-library/react';

import { ColumnLayout } from './ColumnLayout';

function renderLayout() {
  return render(
    <ColumnLayout>
      <div />
      <div />
      <div />
    </ColumnLayout>,
  );
}

describe('<ColumnLayout />', () => {
  it('should match snapshot', () => {
    const component = renderLayout();
    expect(component.asFragment()).toMatchSnapshot();
  });
});
