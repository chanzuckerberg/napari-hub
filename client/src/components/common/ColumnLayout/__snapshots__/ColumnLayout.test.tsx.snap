// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColumnLayout /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 justify-center gap-6 md:gap-12 2xl:grid-cols-napari-2-col 3xl:grid-cols-napari-3-col"
  >
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;
